File: sample_source_code/python/src/ivisitor.py
"""Visitor interface for operations on animal types."""
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from dog import Dog
    from cat import Cat


class IVisitor(ABC):
    """Visitor pattern interface enabling double-dispatch."""

    @abstractmethod
    def visit_dog(self, dog: "Dog") -> None: ...

    @abstractmethod
    def visit_cat(self, cat: "Cat") -> None: ...

File: sample_source_code/javascript/src/IVisitor.js
/**
 * @interface IVisitor
 * Visitor interface for double-dispatch operations on animal types.
 */
class IVisitor {
    /** @param {import('./Dog')} dog */ visitDog(dog) { throw new Error('Not implemented'); }
    /** @param {import('./Cat')} cat */ visitCat(cat) { throw new Error('Not implemented'); }
}

module.exports = IVisitor;
